hips/hip-9999.md

```diff
@@ -0,0 +1,72 @@
+authors: [ "Ralf Van Dorsselaer" ]
+created: "2022-06-20"
+type: "feature"
+status: "draft"
+---
+ 
+## Abstract
+ 
+Helm release data encoded into the release field in a Kubernetes Secret can cause the Secret as a whole to exceed 1 MiB (as defined by MaxSecretSize in kubernetes/pkg/apis/core/types.go).
+ 
+When this happens, the Kubernetes Secret cannot be stored in Etcd and the Helm install or upgrade fails.
+ 
+Splitting the Helm release data across multiple Kubernetes Secrets resolves this problem.
+ 
+## Motivation
+ 
+Helm release data that causes the Kubernetes Secret size to exceed MaxSecretSize will make Helm install or upgrade fail.
+ 
+The existing HELM_DRIVER=secret storage driver does not handle Kubernetes Secrets become too big when storing a large encoded Helm release data.
+ 
+## Rationale
+ 
+The logical place to support this is in the HELM_DRIVER=secret storage driver.
+ 
+## Specification
+ 
+The feature is transparent to users. 
+ 
+When the HELM_DRIVER=secret storage driver notices that the resulting Kubernetes Secret will exceed MaxSecretSize then the storage driver will split the encoded release data across multiple Kubernetes Secrets.
+ 
+Splitting requires additional metadata to be added to the Kubernetes Secret:
+ 
+- number of chunks: integer, the number of Kubernetes Secrets that holds the release data for this Helm release version.
+- chunk: integer, the chunk number. Each Secret gets a chunk number. Reconstituting the release data can be done by reading the Secret with chunk=1, then subsequently reading in numerical order any number of additional Secrets up to and including the "number of chunks".
+ 
+Splitting requires the name of the Secret to include a "chunk" suffix. Example: sh.helm.release.v1.test.v1.2, 3, 4 and so on where the number after the last dot is the "number of chunk". The first Secret will not get a .1 suffix.
+ 
+If no splitting is required then no additional metadata is included/set.
+ 
+## Backwards compatibility
+ 
+If the chunking is implemented in the secret/secrets HELM_DRIVER then backward compatibility to previously installed Helm releases is guaranteed. If no chunking information is encoded in the release secret then the driver can behave as before chunking support was added. If no chunking is needed then the release secret can be created without any additional metadata and with compatible name.
+ 
+## Security implications
+ 
+Theoretically a MTM Man in The Middle attack is possible where requests to the Kubernetes API server are intercepted. But this type of attack applies to any access to the Kubernetes API server.
+ 
+## How to teach this
+ 
+The feature is transparent to users.
+ 
+The documentation (in advanced section(s)) needs to explain that multiple Kubernetes Secrets can be created for a Helm release to counter the size limitation.
+ 
+## Reference implementation
+ 
+PR: https://github.com/helm/helm/pull/11087
+ 
+## Rejected ideas
+ 
+n/a
+ 
```

Comment:
what about changing the compression algorithm ? 
I have tested with one Helm secret that is 284K gzip compressed. Decompressed it is 2.0M
* with bzip2 179K 37% savings
* with brotli 163K 42% savings

Reply:
Valid consideration,  @carlossg.  Compared with the original proposal to store the release in chunked secrets, I'd prefer the chunked storage from the original proposal, for the following reasons:

- chunked storage has no upper limit (for realistic charts - I ignore the theoretical 2GB limit of etcd), whereas a 40% compression gain is IMHO too close to the current limit, i.e. the risk remains that a chart that would need double the size is already out of the picture
- the chunked storage can be backwards-compatible at least in the way that if the release fits into a single secret, the storage can behave exactly as before, and only large charts that could not be installed would need the new version, whereas a compression change would impact all stored releases.

@Hammond95 , the chunked storage is implemented in https://github.com/helm/helm/pull/12277 and pending review.

Reply:
I prefer the compression :) 

- backwards compatibility: you could do the same check, if size under limit use gzip, else use brotli
- forward compatibility: if secret cannot be decompressed with brotli, use gzip
- but the main reason, with chunked storage you still have the data in etcd and etcd has storage limits. helm charts account to a big portion of our etcd storage and anyone doing 100s/1000s of helm releases will have this problem

Reply:
@carlossg , I believe our difference in preference comes from trying to solve different problems.  

This HIP / PR is addressing the limitations that a single release of a large application (== helm chart) can exceed the size limit of the secret storage.  A ~40% increase only shifts the limit of what "too large chart" means, but strategically, I wouldn't be comfortable that this addresses the conceptual problem of stroring in a single secret.

You seem to face rather the problem that you have MANY releases of helm charts, that individually don't run into the 1MB limit of etcd, but in total sum up to exceed etcd overall storage limits, and I'd assume you'd ALWAYS want the improved compression, not only for releases that run into the 1MB limit.  That's a valid problem, and I agree that improved compression can move the needle somewhat as long as the underlying release encoding is not optimized by itself - but as said, IMHO this is not the problem that's in teh focus of this HIP.